### PR TITLE
Support for ubuntu 18.04

### PR DESCRIPTION
**Pull request by kowh-ai**
Added support for ubuntu 18.04 in Vagrantfile, build scripts and ansible.

Tested it on an Ubuntu 18.04 virtual machine with the dev-v2.8 package install instructions and it did work without issues. 

I can make the .deb file available for any testing

I'm more familiar with bitbucket (and CVS from a long time ago :-) ) so please let me know if I'm not following protocols...

### CODE REVIEW

File: package.yml [hunk 1]
```diff
@@ -31,6 +32,10 @@
     - name: stop apache before installing nginx (xenial)
       service: name=apache2 state=stopped
       when: ansible_distribution_version == "16.04"
+      
+    - name: stop apache before installing nginx (bionic)
+      service: name=apache2 state=stopped
+      when: ansible_distribution_version == "18.04"
```
Comment (amercader) [inline package.yml]: Can we make this an OR in the current clause for xenial?

File: package.yml [hunk 2]
```diff
@@ -42,6 +47,10 @@
     - name: install fpm (xenial)
       action: command gem install -v 1.11.0 fpm -- creates=/usr/local/bin/fpm
       when: ansible_distribution_version == "16.04"
+      
+    - name: install fpm (bionic)
+      action: command gem install -v 1.11.0 fpm -- creates=/usr/local/bin/fpm
+      when: ansible_distribution_version == "18.04"
```
Comment (amercader) [inline package.yml]: Same as above

File: ckan-package [hunk 3]
```diff
@@ -40,9 +41,14 @@ def run(target=None):
     if target is None or target == 'xenial':
             status_xenial = re.search('xenial\s*(\w+)', out).group(1)
             print 'Machine "xenial" is ' + status_xenial
+            
+    if target is None or target == 'bionic':
+            status_bionic = re.search('bionic\s*(\w+)', out).group(1)
+            print 'Machine "bionic" is ' + status_bionicl
```
Comment (amercader) [inline ckan-package]: Typo
```suggestion
            print 'Machine "bionic" is ' + status_bionic
```

File: Vagrantfile [hunk 4]
```diff
@@ -13,6 +13,10 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
   config.vm.define "xenial" do |xenial|
     xenial.vm.box = "ubuntu/xenial64"
   end
+  
+  config.vm.define "bionic" do |bionic|
+    xenial.vm.box = "ubuntu/bionic64"
```
Comment (amercader) [inline Vagrantfile]: Typo, that causes an error when running the `ckan-package` command
```suggestion
    bionic.vm.box = "ubuntu/bionic64"
```

File: package.yml [hunk 5]
```diff
@@ -14,6 +14,7 @@
   tasks:
     - name: make sure packages are installed
       apt:
+        force_apt_get: yes
```
Comment (amercader) [inline package.yml]: Is this necessary?

Also I couldn't get the script working without adding `update_cache: yes`

**Comment by amercader** [comment]
I'm getting this error after fixing locally the issues and trying to build a 2.8 bionic package:

```
./ckan-package -v ckan-2.8.3 -i 1 -t bionic
```

```

TASK [install requirements and make virtualenv for ckan] ***********************
task path: /home/adria/dev/pyenvs/ckan-packaging/src/ckan-packaging/package.yml:85
fatal: [bionic]: FAILED! => {"changed": false, "cmd": ["/usr/lib/ckan/default/bin/pip3", "install", "-r", "/usr/lib/ckan/default/src/ckan/requirements.txt"], "msg": "stdout: Already using interpreter /usr/bin/python3\nUsing base prefix '/usr'\nNew python executable in /usr/lib/ckan/default/bin/python3\nAlso creating executable in /usr/lib/ckan/default/bin/python\nInstalling setuptools, pip, wheel...\ndone.\nRunning virtualenv with interpreter /usr/bin/python3\nCollecting argparse==1.4.0 (from -r /usr/lib/ckan/default/src/ckan/requirements.txt (line 7))\n  Downloading https://files.pythonhosted.org/packages/f2/94/3af39d34be01a24a6e65433d19e107099374224905f1e0cc6bbe1fd22a2f/argparse-1.4.0-py2.py3-none-any.whl\nCollecting Babel==2.3.4 (from -r /usr/lib/ckan/default/src/ckan/requirements.txt (line 8))\n  Downloading https://files.pythonhosted.org/packages/b4/ec/acd307eac2e23f9cab1c8bdbe29b3b1d43215e31c32f8aa91b3a97925b5b/Babel-2.3.4-py2.py3-none-any.whl (7.1MB)\nCollecting Beaker==1.9.0 (from -r /usr/lib/ckan/default/src/ckan/requirements.txt (line 9))\n  Downloading https://files.pythonhosted.org/packages/93/b2/12de6937b06e9615dbb3cb3a1c9af17f133f435bdef59f4ad42032b6eb49/Beaker-1.9.0.tar.gz\nCollecting bleach==2.1.3 (from -r /usr/lib/ckan/default/src/ckan/requirements.txt (line 10))\n  Downloading https://files.pythonhosted.org/packages/30/b6/a8cffbb9ab4b62b557c22703163735210e9cd857d533740c64e1467d228e/bleach-2.1.3-py2.py3-none-any.whl\nCollecting click==6.7 (from -r /usr/lib/ckan/default/src/ckan/requirements.txt (line 11))\n  Downloading https://files.pythonhosted.org/packages/34/c1/8806f99713ddb993c5366c362b2f908f18269f8d792aff1abfd700775a77/click-6.7-py2.py3-none-any.whl (71kB)\nCollecting decorator==4.2.1 (from -r /usr/lib/ckan/default/src/ckan/requirements.txt (line 12))\n  Downloading https://files.pythonhosted.org/packages/e1/5a/53db15bf367d2028bdc6700dbdf1bdfab46b9f208b7516952817c0808118/decorator-4.2.1-py2.py3-none-any.whl\nCollecting fanstatic==0.12 (from -r /usr/lib/ckan/default/src/ckan/requirements.txt (line 13))\n  Downloading https://files.pythonhosted.org/packages/b5/93/8b1eb0d99b72799811ebfc76f20480225a3cc0e456452331d7baeb743dd6/fanstatic-0.12.tar.gz (63kB)\n\n:stderr:     ERROR: Command errored out with exit status 1:\n     command: /usr/lib/ckan/default/bin/python3 -c 'import sys, setuptools, tokenize; sys.argv[0] = '\"'\"'/tmp/pip-install-i1oh9fuv/fanstatic/setup.py'\"'\"'; __file__='\"'\"'/tmp/pip-install-i1oh9fuv/fanstatic/setup.py'\"'\"';f=getattr(tokenize, '\"'\"'open'\"'\"', open)(__file__);code=f.read().replace('\"'\"'\\r\\n'\"'\"', '\"'\"'\\n'\"'\"');f.close();exec(compile(code, __file__, '\"'\"'exec'\"'\"'))' egg_info --egg-base pip-egg-info\n         cwd: /tmp/pip-install-i1oh9fuv/fanstatic/\n    Complete output (7 lines):\n    Traceback (most recent call last):\n      File \"<string>\", line 1, in <module>\n      File \"/tmp/pip-install-i1oh9fuv/fanstatic/setup.py\", line 6, in <module>\n        open('CHANGES.txt').read())\n      File \"/usr/lib/ckan/default/lib/python3.6/encodings/ascii.py\", line 26, in decode\n        return codecs.ascii_decode(input, self.errors)[0]\n    UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 293: ordinal not in range(128)\n    ----------------------------------------\nERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.\n"}

```

File: package.yml (hunk 5 above)
Comment (kowh-ai) [reply package.yml]: I built the .deb package on Mac OS X which doesn't have aptitude installed (as far as I know no port for it exists on Mac OS X). I did try to build the .deb package on multiple versions of Ubuntu (16.04, 18.04) but had problems with Vagrant talking to VirtualBox. It was easier to force using apt-get. Whats the "standard" OS for building these .deb packages? Ubuntu version <whatever>? I can revisit getting this working on Ubuntu if you think I should. Also I didn't have to add "update_cache: yes" but it won't hurt to use this option...

File: package.yml (hunk 2 above)
Comment (kowh-ai) [reply package.yml]: Yes - that makes it cleaner - done

File: package.yml (hunk 5 above)
Comment (amercader) [reply package.yml]: @kowh-ai Sorry maybe I miss something but the point of using VirtualBox and Vagrant is that the packages are build in a reproducible environment (the target of the distribution) so the host OS should not matter. But anyway, if `force_apt_get` works better let's keep (along `update_cache: yes`)

**Comment by amercader** [comment]
@kowh-ai Making progress, your latest changes got me past the previous error, now I'm getting this:

```
TASK [copy all distribution specific files] ************************************
task path: /home/adria/dev/pyenvs/ckan-packaging/src/ckan-packaging/package.yml:116
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: If you are using a module and expect the file to exist on the remote, see the remote_src option
failed: [bionic] (item=etc/apache2/sites-available/ckan_default.conf) => {"ansible_loop_var": "item", "changed": false, "item": "etc/apache2/sites-available/ckan_default.conf", "msg": "Could not find or access 'bionic/etc/apache2/sites-available/ckan_default.conf'\nSearched in:\n\t/home/adria/dev/pyenvs/ckan-packaging/src/ckan-packaging/files/bionic/etc/apache2/sites-available/ckan_default.conf\n\t/home/adria/dev/pyenvs/ckan-packaging/src/ckan-packaging/bionic/etc/apache2/sites-available/ckan_default.conf\n\t/home/adria/dev/pyenvs/ckan-packaging/src/ckan-packaging/files/bionic/etc/apache2/sites-available/ckan_default.conf\n\t/home/adria/dev/pyenvs/ckan-packaging/src/ckan-packaging/bionic/etc/apache2/sites-available/ckan_default.conf on the Ansible Controller.\nIf you are using a module and expect the file to exist on the remote, see the remote_src option"}
```

Which makes sense because ansible is looking for a `bionic` folder in the repo that is not there. Maybe you forgot to add it to a commit (it can probably be a copy of the `xenial` one).
There also isn't a `bionic/etc/apache2/sites-available` folder AFAICT.

I did a quick try on a new bionic VM and the datapusher wasn't installed. This is a separate we service with its own Apache config. It is not getting installed because this file also needs updating to support 18.04:

https://github.com/ckan/ckan-packaging/blob/master/common/tmp/after_web.j2

Do a quick audit of the other files in common to see if there is something version specific.

File: package.yml (hunk 5 above)
Comment (kowh-ai) [reply package.yml]: @amercader Agreed that building the package should be done on the target distribution however Iv'e tried may things and haven't got it to work. I was talking to Sergey about this also a couple of days ago and he had the same problem building packages on Ubuntu 18.04. Mac OS X doesn't have an aptitude port AFAIK, hence needed to force "apt-get"

**Comment by kowh-ai** [comment]
Hey @amercader I have added the 2 new files (**after_web.j2** and **ckan_default.conf** and in the correct path)...apologies for missing these out...GitHub is new to me and I'm learning a lot - thanks for your patience :-)

**Comment by kowh-ai** [comment]
Hey @amercader  - have you had a chance to test this again after I made the  changes to after_web.j2 and ckan_default.conf? - thanks

**Comment by kowh-ai** [comment]
@amercader I'm wondering if this can be merged BEFORE I start building packages for hotfix releases: 2.6.9, 2.7.7 and 2.8.4....thanks

**Comment by amercader** [comment]
@kowh-ai I'll do some final testing with this PR and merge it soon

**Comment by amercader** [comment]
@kowh-ai All good! I added a few tweaks to make it work with the latest 18.04 versions. The package builds fine and I did a quick install on a fresh bionic vm and all main features worked.

One thing that needs doing is updating these two pages to reflect that you can do a package install on Ubuntu 18.04:

* https://docs.ckan.org/en/latest/maintaining/installing/index.html
* https://docs.ckan.org/en/latest/maintaining/installing/install-from-package.html

Can you take care of this? Thanks!

**Comment by kowh-ai** [comment]
@amercader Thanks! and yes I can take care of the doco mentioned

**Pull request merged**